### PR TITLE
4400 - Move Camera Button

### PR DESCRIPTION
**Pull request by Cattlesquat**
#4400 You make a button on your map's toolbar. You specify a "destination" the same way you would for a Send To Location trait. It centers the map's camera on that destination. Since it uses expressions and property matches, including the new "cycling property match", you can use it to e.g. cycle the view around "to all the artillery units, one by one".

image

### CODE REVIEW

File: vassal-app/src/main/java/VASSAL/build/module/Map.java [hunk 1]
```diff
@@ -3063,6 +3064,21 @@ public void centerAt(Point p, int dx, int dy) {
...
+    return theMap.getVisibleRect();
+  }
+
+  /**
+   * @return the point our map's view is currently centered on
+   */
+  public Point getCenter() {
+    final Rectangle r = getVisibleRect();
+    return new Point(r.x + r.width/2, r.y + r.height/2);
+  }
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/Map.java:3080]: I'm surprised there wasn't something to do this already.

File: vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+  public MoveCameraButton() {
+    setNameKey("");
+
+    setLaunchButton(makeLaunchButton(
+      Resources.getString("Editor.MoveCameraButton.move_camera_tooltip"),
+      Resources.getString("Editor.MoveCameraButton.move_camera"),
+      "/images/recenter.gif", //NON-NLS
+      e -> {
+        moveCamera();
+      }
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java:86]: One liners you can do like this: `e -> moveCamera()`

File: vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    catch (NumberFormatException e) {
+      ErrorDialog.dataWarning(new BadDataReport(this, Resources.getString("Error.non_number_error"), xOffset.debugInfo(dxString, "X Offset"))); //NON-NLS
+      return;
+    }
+
+    try {
+      dy = Integer.parseInt(dyString);
+    }
+    catch (NumberFormatException e) {
+      ErrorDialog.dataWarning(new BadDataReport(this, Resources.getString("Error.non_number_error"), xOffset.debugInfo(dyString, "Y Offset"))); //NON-NLS
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java:108]: Do you want `yOffset` here instead/

File: vassal-app/src/main/java/VASSAL/build/module/Map.java (hunk 1 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/module/Map.java:3080]: Haha me too, cause it's one of the first things I write in a new videogame!

File: vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java (hunk 2 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java:86]: Yeah I probably deleted some other crap

File: vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java (hunk 3 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java:108]: Oopsie Poopsie?

Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/module/map/MoveCameraButton.java]: fixed

**Pull request merged**